org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest16.java

```diff
@@ -540,6 +540,223 @@ public void doNotMatchBitWiseOr(boolean useStrikethroughForCompleted, Object dat
+		IPackageFragment pack= fSourceFolder.createPackageFragment("test1", false, null);
+		String given= """
+			package test1;
+
+			public class E {
+				public void foo(Object x, boolean valid) {
+					if (x instanceof Integer) {
+						int k = 7;
+						if (valid) {
+							Integer i = (Integer)x;
+							i = 7;
+							System.out.println(i + k);
+						}
+						Integer i = (Integer)x;
+						System.out.println(i);
+					}
+				}
+				public void foo2(Object x, boolean valid) {
+					if (x instanceof Integer) {
+						int k = 7;
+						if (valid) {
+							Integer i = (Integer)x;
+							i = 7;
+							System.out.println(i + k);
+						} else {
+							Integer i = (Integer)x;
+							i = 9;
+							System.out.println(i + k);
+						}
+						Integer i = (Integer)x;
+						System.out.println(i);
+					}
+				}
+				public void foo3(Object x, boolean valid) {
+					if (x instanceof StringBuilder) {
+						int k = 7;
+						if (valid) {
+							StringBuilder s = (StringBuilder)x;
+							s.append('A');
+							System.out.println(s.length() + k);
+						} else {
+							StringBuilder s = (StringBuilder)x;
+							s.append('B');
+							System.out.println(s.length() + k);
+						}
+						StringBuilder s = (StringBuilder)x;
+						System.out.println(s);
+					}
+				}
+			}
+			""";
+		ICompilationUnit cu= pack.createCompilationUnit("E.java", given, false, null);
+
+		enable(CleanUpConstants.USE_PATTERN_MATCHING_FOR_INSTANCEOF);
+		String expected= """
+				package test1;
+
+				public class E {
+					public void foo(Object x, boolean valid) {
+						if (x instanceof Integer i) {
+							int k = 7;
+							if (valid) {
+								i = 7;
+								System.out.println(i + k);
+							}
+							i = (Integer) x;
+							System.out.println(i);
+						}
+					}
+					public void foo2(Object x, boolean valid) {
+						if (x instanceof Integer i) {
+							int k = 7;
+							if (valid) {
+								i = 7;
+								System.out.println(i + k);
+							} else {
+								i = 9;
+								System.out.println(i + k);
+							}
+							i = (Integer) x;
+							System.out.println(i);
+						}
+					}
+					public void foo3(Object x, boolean valid) {
+						if (x instanceof StringBuilder s) {
+							int k = 7;
+							if (valid) {
+								s.append('A');
+								System.out.println(s.length() + k);
+							} else {
+								s.append('B');
+								System.out.println(s.length() + k);
+							}
+							System.out.println(s);
+						}
+					}
+				}
+				""";
+
+		assertNotEquals("The class must be changed", expected, given);
+		assertRefactoringResultAsExpected(new ICompilationUnit[] { cu }, new String[] { expected },
+				new HashSet<>(Arrays.asList(MultiFixMessages.PatternMatchingForInstanceofCleanup_description)));
+	}
+
+	@Test
+	public void testDeeperMatches() throws Exception {
+		IPackageFragment pack= fSourceFolder.createPackageFragment("test1", false, null);
+		String given= """
+			package test1;
+
+			public class E {
```

Comment:
Please add a scenario where you have 2 instanceofs separated by and.  This PR fixes another bug where 2 instanceofs can be converted in one cleanup and your fix handles it.

e.g. 
```
	public void foo(Object o, Object p) {
		if (o instanceof String && p instanceof Integer) {
			if (o != p) {
				String s = (String) o;
				Integer i = (Integer) p;
				System.out.println(s.trim() + i.toString());
				i = 3;
			}
			Integer i = (Integer) p;
			i = 7;
		}
	}
```

Reply:
I'm really surprised, but yes, the code handles this too.